lib/endpoints/class-wp-rest-posts-controller.php

```diff
@@ -1874,4 +1876,73 @@ public function validate_user_can_query_private_statuses( $value, $request, $par
 		}
 		return new WP_Error( 'rest_forbidden_status', __( 'Status is forbidden' ), array( 'status' => rest_authorization_required_code() ) );
 	}
+
+	/**
+	 * Validate filter param.
+	 *
+	 * This function can be used to validate query params passed by the filter request param.
+	 *
+	 * @param  mixed           $value     Value of the parameter.
+	 * @param  WP_REST_Request $request   Request body.
+	 * @param  string          $parameter Name of the parameter. Should be tax_query.
+	 * @return WP_Error|boolean
+	 */
+	public function rest_validate_query_filter( $values, $request, $parameter ) {
+		// Returns true for validation by default if 'tax_query' is not set in the filter.
+		$is_valid = true;
+		if ( isset( $values['tax_query'] ) ) {
+			$is_valid = $this->rest_validate_tax_query( $values['tax_query'], $request, 'tax_query' );
+			// If an error is found exit early and it will throw an error for request->has_valid_params();
+			if ( is_wp_error( $is_valid ) ) {
+				return $is_valid;
+			}
+		}
+		return $is_valid;
+	}
+
+	/**
+	 * Validate tax query.
+	 *
+	 * @param  mixed           $value      Value of the parameter.
+	 * @param  WP_REST_Request $request    Request body.
+	 * @param  string          $parameter  Name of the parameter. Should be tax_query.
+	 * @param  array           $taxonomies Taxonomies that are to be shown in the REST API.
+	 * @return WP_Error|boolean
+	 */
+	public function rest_validate_tax_query( $values, $request, $parameter, $taxonomies = array() ) {
+		if ( 'tax_query' !== $parameter ) {
+			return new WP_Error( 'rest_invalid_param', __( 'This validation method is only used to validate a tax query.' ), array( 'status' => 400 ) );
+		}
+		$valid = true;
+		if ( ! empty( $values ) ) {
+			foreach ( $values as $key => $value ) {
+				// Switch statement to handle different cases.
+				if ( is_array( $value ) && 'terms' !== $key ) {
+					// Recursive call to function to handle nested taxonomy queries.
+					$valid = $this->rest_validate_tax_query( $value, $request, $parameter, $taxonomies );
+
+					// If an error is found keep returning it up the chain.
+					if ( is_wp_error( $valid ) ) {
+						return $valid;
+					}
+				}
+
+				if ( 'taxonomy' === $key ) {
+					// If $taxonomies is empty by default set it. On reoccurring calls the function overhead is avoided.
+					if ( empty( $taxonomies ) ) {
+						$taxonomies = wp_list_filter( get_object_taxonomies( $this->post_type, 'objects' ), array( 'show_in_rest' => true ) );
```

Comment:
Could/should this check & assignment go at the top of this block, before the `foreach`? I doubt there's much perf impact here but since the list of taxonomies that apply to this request's post type won't be changing during this loop, the list of `$taxonomies` could be declared outside the foreach loop.